python/ray/serve/config.py

```diff
@@ -198,8 +201,23 @@ def _validate(self):
             self.resource_dict.update(custom_resources)
 
 
+class DeploymentMode(Enum, str):
+    NoServer = "NoServer"
+    NodePort = "NodePort"
+    HeadOnly = "HeadOnly"
+
+
 @dataclass
-class HTTPConfig:
-    host: str = field(init=True)
-    port: int = field(init=True)
-    middlewares: List[Any] = field(init=True)
+class HTTPOptions:
+    host: Union[None, str] = DEFAULT_HTTP_HOST
+    port: int = DEFAULT_HTTP_PORT
+    middlewares: List[Any] = field(default_factory=list)
+    mode: Union[None, DeploymentMode] = DeploymentMode.HeadOnly
+    _head_node_id: Optional[ray._raylet.NodeID] = None
+
+    def __post_init__(self):
+        if self.mode is None or self.host is None:
+            self.mode = DeploymentMode.NoServer
+
+        if self._head_node_id is None:
+            self._head_node_id = ray.get_runtime_context().node_id
```

Comment:
Can we validate the mode here? Or is it already dynamically checked?

Reply:
One way to do the checks dynamically is to use a pydantic BaseModel, or a pydantic dataclass.  Could be worth it, but I'm not sure